### ISSUE TITLE
On QNX ptpd only works when RUNTIME_DEBUG is enabled on master

### ISSUE THREAD

**Issue by da-phil** [ptpd/ptpd#4]
I'm currently testing the latest version of ptpd on QNX 6.5.0 machines and apparently neither multicast nor unicast communication work, the ptpd nodes just don't 'find' each other.
However, during debugging the issue I found it really weird that once I build ptpd with RUNTIME_DEBUG enabled:
```
./configure --enable-runtime-debug
```
and turning it on by adding ```-D``` to the invocation on the master side it somehow works.

I've encountered many ```#ifdef RUNTIME_DEBUG``` statements everywhere in the code, so I think somewhere in combination with a ```#ifdef __QNXNTO__``` statement might be the bug.

**Comment by wowczarek** [comment]
Hi,

I saw a similar problem on 6.5.0 as well, but I took it for NIC issues. Are you running on physical hardware? In my case running in sniffing more (ptpengine:use_libpcap=y) or using unicast solved the issue. Or running tcpdump simultaneously. I take it you're using multicast?

I need to get QNX rocking on some dedicated hardware to test this again. If running in debug mode fixes this then it makes this more interesting. Running on a VM emulating Intel e1000 I never see this issue - running on physical Intel 8257x I do.

**Comment by da-phil** [comment]
Hi!

Yep, I'm running on physical hardware and tried both, multicast and unicast, both was affected. But I didn't try sniffing mode so far, I'll try it next week.

**Comment by wowczarek** [comment]
OK, good to have somebody else testing it on QNX. Just remember that you basically need to compile with ```--enable-experimental-options``` on QNX as only then you will get any meaningful timing quality. PTPd attaches itself to the timer interrupt and makes heavy use of it for reading time (it uses the TSC counter delta since last timer interrupt to approximate time). I decided to have this under experimental, because it calls ```clock_gettime()``` inside an interrupt handler, to which QNX manuals say "no". So when you enable experimental, it will read time when receiving packets and sending packets, instead of using socket timestamps, which have really poor granularity. This method works fine because of QNX's high determinism. I did not invent this as this came from earlier patches, but I refined it.

**Comment by da-phil** [comment]
Now I used ```--enable-experimental-options``` and activated sniffing mode (```ptpengine:use_libpcap=y```) but it seems that there is a problem. The slave appears to do nothing, it's neither showing the estimated network delay nor the clock offset.
I attached two pairs of logs
- master & slave output in multicast mode
- master & slave output in unicast mode

You can find the command-line call with all used parameters in the first line in every log file.

[ptpd-master-multicast01.txt](https://github.com/ptpd/ptpd/files/47479/ptpd-master-multicast01.txt)
[ptpd-master-unicast01.txt](https://github.com/ptpd/ptpd/files/47478/ptpd-master-unicast01.txt)
[ptpd-slave-multicast01.txt](https://github.com/ptpd/ptpd/files/47477/ptpd-slave-multicast01.txt)
[ptpd-slave-unicast01.txt](https://github.com/ptpd/ptpd/files/47476/ptpd-slave-unicast01.txt)

**Comment by wowczarek** [comment]
Hi,

It is definitely receiving all the required message types for unicast and multicast slave in the two logs you sent, so it's not the network. It is computing the offsets. In unicast mode it seems to keep resetting the clock, in multicast it looks like it's starting to sync. Mind that in the default config it will step the clock when offset is >1s, and then reset the state machine. It will also not use the mean path delay if it is negative - which often is the case when there is a very high initial offset. The value becomes usable only some time later.

Based on what output are you saying it is not showing the offsets / delays? The fact itself that the slave is in slave state means it's at least receiving Announce.

Check the man pages (```ptpd2.8```, ```ptpd2.conf.5```) for other ways of displaying the slave status, namely the status file which gives you a single-screen overview of the slave's state.

Multicast:

```2015-11-30 15:41:55.200316, slv, 180373fffe2ddcc0(unknown)/1,  0.000000000, -0.579423134,  0.579423135, -0.579423135, -500000.000000000, S, 00021, 0.000000000, 0, 0.000000000, 0, 0, 0, -0.579423135,  0.579423135```

Unicast:

```2015-11-30 15:57:40.833772, slv, 180373fffe2ddcc0(unknown)/1,  0.000000000,  0.000000000,  0.000000000,  0.000000000, 0.000000000, I, 00514, 0.000000000, 0, 0.000000000, 0, 0, 0, -31.583016210,  0.000000000```

Also looks like you have a very high Announce interval set, like 16 per second - is this intentional?

**Comment by da-phil** [comment]
The high announce frequency was chosen intentionally because I wanted to make the waiting time until synchronization as short as possible. I've tried various values and it doesn't seem to cause any problem using higher frequencies.

I really don't understand whats going on, I'm currently testing with commit 1f0baae98a7b23e85f2bfd8f5de64795421c270e (- Added QNX support), to make sure there was no regression between master and this commit.

This version doesn't work most of the time, see output here:
[ptpd-master-multicast03.txt](https://github.com/ptpd/ptpd/files/50034/ptpd-master-multicast03.txt)
[ptpd-slave-multicast03.txt](https://github.com/ptpd/ptpd/files/50035/ptpd-slave-multicast03.txt)
the slave always ends up in a PTP_FAULTY state.

I somehow managed to get it working once, however I can't reproduce, so I don't know why it worked.

Only your patch for version ptpd-2.3.1-rc1 (https://sourceforge.net/p/ptpd/patches/57) seems to work for QNX.

**Comment by wowczarek** [comment]
That's fine re. announce interval - I just wasn't sure why you used a high interval. When you say that only the patch works for QNX, you mean it only works for QNX in your case. I did not commit the code without testing it. I tested on an x86 VM and x86 hardware and did not see the issues.

Also the slave does not end up in FAULTY state, it only goes through the faulty state (it does it after clock step to ensure full state machine reset). You can clearly see in the log that it's constantly seeing offsets of multiple seconds and it keeps reseting the clock. This is not a communication issue. The issue I said I was seeing was where the process would not receive any data - which is not the case here, so this is unrelated.

So you're saying that it works every time with the old patch? What is the exact hardware you are using, what CPU architecture? What NIC?

**Comment by da-phil** [comment]
Ok, now that I've played again with the announce interval I have to say that my previous statement above was totally wrong, the announce interval plays an important role, setting it to a smaller value actually made ptpd sync work.
Apparently I only tested values <=0, which means announce rates of >=1Hz where synchronization seemed to work properly (but at that time I didn't know about the nex step of turning on runtime debug...).
The default value for ptpengine:log_announce_interval is 2s (1/2 Hz announce rate) and this value let the previous test fail, because I didn't change that value to achieve a faster announce rate.

I can also confirm that the behaviour is the same with the latest commits on master branch, so there seems to be no regression in the QNX case.

Now I changed back to the head of master branch and created two more logs, both with a lot of debug output (option `-DDD`) to show whats going on and maybe even see why it fails to sync in the first case:
- explicit setting for the default announce interval, which is 2^1 = 2s  (`--ptpengine:log_announce_interval=1`), where syncing **didn't work**
  - master> `ptpd2 -m -i wm0 -V -DDD --ptpengine:log_announce_interval=0` [(logfile)](https://github.com/ptpd/ptpd/files/50779/ptpd-master-multicast04.txt)
  - slave> `ptpd2 -m -i wm0 -V -DDD` 
[(logfile)](https://github.com/ptpd/ptpd/files/50780/ptpd-slave-multicast04.txt)
- decreasing announce interval to 2^0 = 1s (`--ptpengine:log_announce_interval=0`) where syncing **worked**
  - master> `ptpd2 -m -i wm0 -V -DDD --ptpengine:log_announce_interval=0` [(logfile)](https://github.com/ptpd/ptpd/files/50836/ptpd-master-multicast05.txt)
  - slave> `ptpd2 -m -i wm0 -V -DDD` [(logfile)](https://github.com/ptpd/ptpd/files/50837/ptpd-slave-multicast05.txt)

So to sum it up, in order to make ptpd sync properly for QNX systems I needed to do three things:
- Activate runtime debug option during compile time: `./configure --enable-runtime-debug`
- Activate debug messages on ptpd master during runtime by adding `-D` to the commandline call
- Decrease announce time to <=1s, e.g. `--ptpengine:log_announce_interval=0`

Wow, now it took me a while to figure out a working configuration, do you have any idea what happens there? I didn't have enough time to interpret and understand the logs.

**Comment by da-phil** [comment]
I just added the hardware details you were asking for in my post above:
For the test on QNX 6.5.0 I was using identical off the shelf Dell Precision workstations, with Intel Sandy Bridge chipset, core-i processor and intel gigabit 82574L NIC.

I can't imagine that this issue is related to hardware or drivers.

**Comment by morteza1984** [comment]
Dear Wojciech and developers, 
Thanks for the nice job! 
I downloaded the qnx port 1f0baae (- Added QNX support) and wanted to run it on an ARM processor (Beaglebone Black with am3358 from TI) as slave. Without --enable-experimental-options, I get it compiled and I see very low-performance synchronization values. With --enable-experimental-options, it compiles but the offset get strange values and the clock steps all the time. I really apprciate any help. I attached the slave output.
Thanks and best regards, 
Morteza
[experimental-slave.txt](https://github.com/ptpd/ptpd/files/1968668/experimental-slave.txt)
[non-experimental-slave.txt](https://github.com/ptpd/ptpd/files/1968669/non-experimental-slave.txt)
